MatrixSDK/Crypto/CrossSigning/MXCrossSigning.m

```diff
@@ -282,12 +291,51 @@ - (void)requestPrivateKeysToDeviceIds:(nullable NSArray<NSString*>*)deviceIds
 {
     NSLog(@"[MXCrossSigning] requestPrivateKeysToDeviceIds: %@", deviceIds);
     
-    // Make a secret share request for USK and SSK
+    // Make a secret share request for MSK, USK and SSK
     dispatch_group_t successGroup = dispatch_group_create();
     dispatch_group_t onPrivateKeysReceivedGroup = dispatch_group_create();
     
-    __block NSString *uskRequestId, *sskRequestId;
+    __block NSString *mskRequestId, *uskRequestId, *sskRequestId;
+    
+    
+    // MSK
+    dispatch_group_enter(successGroup);
+    dispatch_group_enter(onPrivateKeysReceivedGroup);
+    [self.crypto.secretShareManager requestSecret:MXSecretId.crossSigningMaster toDeviceIds:deviceIds success:^(NSString * _Nonnull requestId) {
+        mskRequestId = requestId;
+        dispatch_group_leave(successGroup);
+    } onSecretReceived:^BOOL(NSString * _Nonnull secret) {
+        
+        BOOL isSecretValid = NO;
+        if (self.myUserCrossSigningKeys.masterKeys.keys)
+        {
+            isSecretValid = [self isSecretValid:secret forPublicKeys:self.myUserCrossSigningKeys.masterKeys.keys];
+        }
+        else
+        {
+            // Accept the secret anyway (It should not happen)
+            isSecretValid = YES;
+        }
+        
+        NSLog(@"[MXCrossSigning] requestPrivateKeysToDeviceIds: Got MSK. isSecretValid: %@", @(isSecretValid));
+        if (isSecretValid)
+        {
+            [self.crypto.store storeSecret:secret withSecretId:MXSecretId.crossSigningMaster];
+            dispatch_group_leave(onPrivateKeysReceivedGroup);
+        }
```

Comment:
We are missing balance call to `dispatch_group_leave(onPrivateKeysReceivedGroup);` when `isSecretValid == NO` for each key.
From `dispatch_group_enter` [documentation](https://developer.apple.com/documentation/dispatch/1452803-dispatch_group_enter): 
> A call to this function must be balanced with a call to dispatch_group_leave.

Reply:
The current algorithm seems correct to me. If the secret is not valid, we need to wait for another device to send another secret.
This blocks return NO to keep receiving secrets.